morf-oracle/src/main/java/org/alfasoftware/morf/jdbc/oracle/OracleDialect.java

```diff
@@ -152,6 +156,96 @@ public Collection<String> internalSequenceDeploymentStatements(Sequence sequence
   }
 
 
+  @Override
+  public List<String> getSchemaConsistencyStatements(SchemaResource schemaResource) {
+    return schemaResource.getAdditionalMetadata()
+        .map(this::getSchemaConsistencyStatements)
+        .orElseGet(() -> super.getSchemaConsistencyStatements(schemaResource));
+  }
+
+
+  private List<String> getSchemaConsistencyStatements(AdditionalMetadata metaDataProvider) {
+    return FluentIterable.from(metaDataProvider.tables())
+        .transformAndConcat(table -> healTable(table, metaDataProvider))
+        .toList(); // turn all the concatenated fluent iterables into a firm immutable list
+  }
+
+
+  private Iterable<String> healTable(Table table, AdditionalMetadata metaDataProvider) {
+    Collection<String> primaryKeyIndexNames = metaDataProvider.primaryKeyIndexNames();
+    Iterable<String> statements = healTruncatedIndexesSequencesAndTriggers(table, primaryKeyIndexNames);
+
+    if (statements.iterator().hasNext()) {
+      List<String> intro = ImmutableList.of(convertCommentToSQL("Auto-Healing table: " + table.getName()));
+      return Iterables.concat(intro, statements);
+    }
+    return ImmutableList.of();
+  }
+
+  private Iterable<String> healTruncatedIndexesSequencesAndTriggers(Table table, Collection<String> primaryKeyIndexNames) {
+    List<String> statements = Lists.newArrayList();
+    // Truncation of indexes will only have occurred for tables with names that are over 27 characters.
+    if (table.getName().length() < 28) return statements;
+
+    // Check if a truncated PK index exists.
+    String truncatedIndexName = truncatedTableNameWithSuffixLegacy(table.getName(), "_PK");
+    if (!primaryKeyIndexNames.stream().anyMatch(indexName -> indexName.equalsIgnoreCase(truncatedIndexName))) return statements;

```

Comment:
Note that this primaryKeyIndexNames is in fact a TreeSet.
Which means using primaryKeyIndexNames.contains() would make this fast.
But then again, that is a very sneaky implementational detail, rather easy to regress in the future.
(And it would only be a TreeSet if coming from Oracle, so this would definitely need to be only from Oracle)

Perhaps a Map would be easier to work with, hopefully from both ends?

Reply:
Let's go with changing AdditionalMetadata to return a map from table name to primary key index name. The alternative would be sorting out the case so contains can be used (most Set implementations should be good with that) and changing the API to return Set rather than Collection. However, the map feels a bit more flexible for the future.